### PR TITLE
Edit device in json feature and add gateway group membership

### PR DESCRIPTION
**Pull request by ajaypratap003**
### Type of change

- Bugfix
- Enhancement / new feature
- Refactoring

### Description
Edit device in json feature and add gateway group membership

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [ ] Write tests and make sure they pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: console/console-init/ui/src/components/TypeAheadSelect/TypeAheadSelect.tsx [hunk 1]
```diff
@@ -79,24 +82,28 @@ const TypeAheadSelect: React.FunctionComponent<ITypeAheadSelectProps> = ({
...
     }
 
-    const options = [
-      <SelectOption
-        value={TypeAheadMessage.MORE_CHAR_REQUIRED}
-        key="1"
-        isDisabled={true}
-      />
-    ];
+    let options: any[] = [];
```
Comment (anukritijha) [inline console/console-init/ui/src/components/TypeAheadSelect/TypeAheadSelect.tsx:97]: Can the type here be SelectOptionObject[]

Comment (ajaypratap003) [reply console/console-init/ui/src/components/TypeAheadSelect/TypeAheadSelect.tsx:97]: I tried to use SelectOptionObject but it cause the issue. based on the situation internal option will change.

File: console/console-init/ui/src/graphql-module/queries/iot_device.ts [hunk 2]
```diff
@@ -21,15 +21,17 @@ const RETURN_IOT_DEVICE_DETAIL = (
         total
         devices{
           deviceId
-          enabled
-          via
-          viaGroups
+          registration{
+            enabled
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/graphql-module/queries/iot_device.ts:25]: memberOf and defaults are missing

File: console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/DeviceInfo.tsx [hunk 3]
```diff
@@ -62,8 +62,10 @@ export const DeviceInfo: React.FC<IDeviceInfoProps> = ({
   const { projectname, namespace } = useParams();
 
   const jsonViewData = {
-    via: deviceList,
-    ...metadetaJson,
+    registration: {
+      via: deviceList,
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device-detail/components/DeviceInfo/DeviceInfo.tsx:66]: viaGroups and memberOf are missing

File: console/console-init/ui/src/modules/iot-device-detail/containers/DeviceInfoContainer/DeviceInfoContainer.tsx [hunk 4]
```diff
@@ -28,10 +28,12 @@ export const DeviceInfoContainer: React.FC<IDeviceInfoContainerProps> = ({
   const { projectname, deviceid, namespace } = useParams();
   const queryResolver = `
     devices{
-      via
-      credentials
-      ext
-      defaults
+      registration{
+        via
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device-detail/containers/DeviceInfoContainer/DeviceInfoContainer.tsx:32]: viaGroups and memberOf are missing

File: console/console-init/ui/src/modules/iot-device/containers/EmptyDeviceContainer/EmptyDeviceContainer.tsx [hunk 5]
```diff
@@ -42,12 +49,24 @@ export const EmptyDeviceContainer: React.FC<IEmptyDeviceContainerProps> = ({
...
   const total = data?.devices?.total || 0;
 
-  total > 0 && setTotalDevices(total);
+  useEffect(() => {
+    if (total === 0) {
+      setShowEmptyDevice(true);
+    } else if (total > 0) {
+      setShowEmptyDevice(false);
+    }
+  }, [total]);
```
Comment (drcoolsanjeev) [inline console/console-init/ui/src/modules/iot-device/containers/EmptyDeviceContainer/EmptyDeviceContainer.tsx:58]: just passing total will create warning. Please resolve this warning if possible

File: console/console-init/ui/mock-console-server/mock-console-server.js [hunk 6]
```diff
@@ -1981,7 +1981,7 @@ function toggleIoTProjectStatus(iotProject, status) {
 }
 
 function toggleIoTDevicesStatus(iotProject, iotDevice, status) {
-  getIotDevice(iotProject, iotDevice).enabled = status;
+  getIotDevice(iotProject, iotDevice).registration.enabled = status;
```
[the remark this answers is not in the document]
Comment (ajaypratap003) [reply console/console-init/ui/mock-console-server/mock-console-server.js:1984]: This is mock-server code not UI. I just added registration as schema has changed.

File: console/console-init/ui/mock-console-server/mock-console-server.js [hunk 7]
```diff
@@ -1981,7 +1981,7 @@ function toggleIoTProjectStatus(iotProject, status) {
...
 }
 
 function getMockIotDownstreamMessagingAddresses() {
```
[the remark this answers is not in the document]
Comment (ajaypratap003) [reply console/console-init/ui/mock-console-server/mock-console-server.js:1987]: This is mock-server code and name decided by JB.

File: console/console-init/ui/src/components/JsonEditor/JsonEditor.tsx [hunk 8]
```diff
@@ -30,10 +30,17 @@ const JsonEditor: React.FunctionComponent<IJsonEditorProps> = ({
...
+
   const onChange = (value: string) => {
-    setDetail && setDetail(value.trim());
+    setValue(value);
   };
-  const [isCopied, setIsCopied] = useState<boolean>(false);
+
+  useEffect(() => {
+    setDetail && setDetail(value);
+  }, [value]);
```
[the remark this answers is not in the document]
Comment (ajaypratap003) [reply console/console-init/ui/src/components/JsonEditor/JsonEditor.tsx:42]: This is kind of common probelms in useEffect so we will have to find out the solution.

File: console/console-init/ui/src/modules/iot-device/containers/EmptyDeviceContainer/EmptyDeviceContainer.tsx (hunk 5 above)
Comment (ajaypratap003) [reply console/console-init/ui/src/modules/iot-device/containers/EmptyDeviceContainer/EmptyDeviceContainer.tsx:58]: This is kind of common probelms in useEffect so we will have to find out the solution.

**Pull request merged**